File: 118-pascals-triangle/pascals-triangle.java
class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> triangle = new ArrayList<>();
        
        for (int i = 0; i < numRows; i++) {
            List<Integer> row = new ArrayList<>();
            for (int j = 0; j <= i; j++) {
                // Every cell in Pascal's triangle is exactly nCr!
                row.add(nCr(i, j));
            }
            triangle.add(row);
        }
        return triangle;
    }
    
    // Helper function to calculate combinations (n choose r)
    private int nCr(int n, int r) {
        long res = 1;
        // Optimize by taking the smaller half
        if (r > n - r) r = n - r; 
        
        for (int i = 0; i < r; i++) {
            res = res * (n - i);
            res = res / (i + 1);
        }
        return (int) res;
    }
}

File: 118-pascals-triangle/pascals-triangle.cpp
class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        vector<vector<int>> ans;
        int n= ans.size();
        

        if(numRows==1){
            ans.push_back({1});
            return ans;
        }

        else{
        ans.push_back({1});
        ans.push_back({1,1});
        for(int i=2; i<numRows; i++){
            vector<int> temp(i+1);
            temp[0]=1;
            
            for(int j=1; j<i; j++){
                temp[j]= ans[i-1][j-1]+ans[i-1][j];
                }
            temp[i]=1;
            ans.push_back(temp);
            
        }
        return ans;
        }

        
    }
};